feat: Proofs for Bits/Fast/FiniteStateMachine, Bits/Fast/Reflect

bollu opened the pull request:
- [x] Prove that the FSMs correctly simulate the bitstream operations
- [x] Prove that the bitstream operations match the denotations
- [x] Put both proofs together to show an end-to-end proof.

bollu committed:
chore: nuke sorrys

bollu committed:
chore: show correctness of all width theorems

bollu committed:
chore: continue proving equivalence of bitstream and FSM operations.

bollu committed:
chore: get predicate FSM eval sorry free

bollu committed:
chore: eliminate more sorrys.

bollu committed:
chore: delete more sorrys

bollu committed:
chore: one more sorry

bollu committed:
chore: cleanup warning

bollu committed:
chore: remove BitStream.denote, just use BitStream.toBitVec

bollu committed:
chore: show that bitstream and bitvec semantics agree for Term

bollu committed:
chore: make a dent on predicate semantics

bollu committed:
chore: prove land / lor

bollu committed:
chore: cleanup

bollu committed:
feat: show that equality predicate is well-formed

bollu committed:
chore: show that not-equals also preserves semantics

bollu committed:
chore: start proving correctness of arithmetic expressions

This needs us to show BitVec and BitStream builds the same circuits,
for which we build the same API.

bollu committed:
chore: showed that add works correctly

bollu committed:
feat: show equivalence for all other ops other than ls

bollu committed:
chore: Term.eval_eq_denote is sorry-free

bollu committed:
chore: drop a ton of useless code:

bollu committed:
chore: simplify dependencies of proof

bollu committed:
chore: accurately keep track of assumptions of sorrys made in decide correctness

tobiasgrosser commented on SSA/Experimental/Bits/Fast/Decide.lean:
```diff
@@ -23,154 +13,41 @@ instance (p : Predicate) :
-#guard_msgs in #eval! decide (or x y) (and x (not y) + y)
-
-/-- info: true -/
-#guard_msgs in #eval! decide (and x y) (or (not x) y - not x)
```
Why do you drop the test cases?

bollu replied:
The tests have been ported to `Bits/Fast/Reflect.lean`, they are not dropped.

This tests the old decision procedure, which tries to verify `t1.eval = t2.eval`. I drop support for the old version, so I can focus on getting the new version sorry free. Maintaining both adds a bit too much friction to the proof process. I'm pretty close to an end-to-end, sorry free proof.

Note that our current framework subsumes this, as we can simulate `t1.eval = t2.eval` with `Predicate.eval (Predicate.eq t1 t2)`.

bollu committed:
chore: brute-force a proof of evalFin eq eval for term/predicate

bollu committed:
chore: make progress on ule. Have ult/slt/sle to do

Annoyingly, we don't seem to have enough theory around slt? wtf.

bollu committed:
chore: remove axioms from eval = evalFin.

also make progress on ule and sle.
TODO: prove the correctness of ult, slt
using:

> BitVec.ult_eq_not_carry
> BitVrc.slt_eq_ult

bollu committed:
chore: cleanup FSM defs

This lets us write the FSM construction modularly,
so that we can easily show that e.g. signed less than is reducible
to unsigned less than plus a branch on the MSB.

bollu committed:
feat: refactor to use BitVec.slt_eq_ult definition

bollu committed:
chore: fixup slt/sle semantics to be what Lean uses

bollu committed:
chore: cleanup slt statement

bollu committed:
chore: show that Slt does the right thing, assuming Ult does

bollu committed:
chore: hard bit done, now the painful bit!

bollu committed:
feat: I win the bet :)

bollu committed:
chore: cleanup
